src/core/query.ts

```diff
@@ -219,14 +217,19 @@ export class Query<
     }
   }
 
-  setData(
-    updater: Updater<TData | undefined, TData>,
-    options?: SetDataOptions
-  ): TData {
-    const prevData = this.state.data
+  setData(data: TData, options?: SetDataOptions): TData {
+    if (typeof data === 'undefined') {
+      const message = 'Cache data cannot be undefined'
+      getLogger().error(message)
+      const error = new Error(message) as any
+      this.dispatch({
+        type: 'error',
+        error,
```

Comment:
my idea was to _not_ do this in `setData`, but keep `setData` to only accept `TData` and do the transformation on the only call-side:

https://github.com/tannerlinsley/react-query/blob/c0fc916cb3e7d32a05e9356b2f1359c3cd5de573/src/core/query.ts#L445-L446

there we could do the check and then:
- extract the `onError` function of the retrier to it's own function
- just call `onError` with the newly created error. That would:
  - dispatch to error state
  - already does the logging in dev mode for us:
https://github.com/tannerlinsley/react-query/blob/c0fc916cb3e7d32a05e9356b2f1359c3cd5de573/src/core/query.ts#L470-L472

Reply:
Sounds way better On it.